test/default_field_linter.lean

```diff
@@ -0,0 +1,66 @@
+import Mathlib.Algebra.Ring.Defs
+
+def Foo := Unit
+
+
+/--
+error: While evaluating default nsmul := nsmulRec:
+  failed to synthesize instance
+    Zero Foo
+---
+error: While evaluating default zsmul := zsmulRec:
+  failed to synthesize instance
+    Zero Foo
+-/
+#guard_msgs in
+instance : Ring Foo where
+  add_assoc _ _ _ := rfl
+  zero := ()
+  zero_add _ := rfl
+  add_zero _ := rfl
+  add_comm _ _ := rfl
+  mul _ _ := ()
+  left_distrib _ _ _ := rfl
+  right_distrib _ _ _ := rfl
+  zero_mul _ := rfl
+  mul_zero _ := rfl
+  mul_assoc _ _ _:= rfl
+  one := ()
+  one_mul _ := rfl
+  mul_one _ := rfl
+  add_left_neg _ := rfl
+
+-- this is the fix to the errors above
```

Comment:
This was not needed when we have default values, right?

Reply:
Correct: but writing things this way is maybe a good thing to teach anyway, as otherwise you don't have anywhere to put your `rfl` lemmas saying what the operation does. This approach also resulted in quite a large performance boost in #6262 over trying to inline everything.

Reply:
Can you modify the `exact_with_diamond_warning` so that if no instance for the required class is found, it will use the value of the field with the corresponding name?.

I think the behavior after this PR is still really bad for new users (and for teaching new users). It's hardly better than the status quo, and much worse than

I don't even know why #6262 is done. The PR description lists one reason, which seems incredibly weak to me.

> The motivation here is to make it more obvious when module diamonds are being introduced, or at least where they might be hiding; you can now simply search for nsmulRec in the source code.

I agree that when constructing new groups from old ones, you should not use the default value, and I'm happy to have a linter for that in Mathlib, but not at the cost of being unable to define what a group is with beginners without explaining what `nsmulRec` is and why we even need `nsmul` as a field.

Feel free to respond to these questions on Zulip, I believe no discussion about this happened before #6262 was merged.

Reply:
Do you know why these errors happen? Do they happen when writing `nsmul := by exact nsmulRec`? Are you giving the expected type to the `exact` tactic?

Reply:
I believe these errors happen because `foo := bar` defaults have special elaboration behavior that change the order in which the constructor arguments are elaborated, while `foo := by exact bar` does the more standard "meta variable populated by tactic" thing.